File: Answer/2943.最大化网格图中正方形空洞的面积.py
#
# @lc app=leetcode.cn id=2943 lang=python3
#
# [2943] 最大化网格图中正方形空洞的面积
#

# @lc code=start
from typing import List


class Solution:
    def maximizeSquareHoleArea(self, n: int, m: int, hBars: List[int], vBars: List[int]) -> int:
        hBars.sort()
        hMax = 1
        left = 1
        right = 2
        for bar in hBars:
            if bar == right:
                right = bar + 1
            else:
                left = bar - 1
                right = bar + 1
            hMax = max(hMax, right - left)

        vBars.sort()
        vMax = 1
        left = 1
        right = 2
        for bar in vBars:
            if bar == right:
                right = bar + 1
            else:
                left = bar - 1
                right = bar + 1
            vMax = max(vMax, right - left);
        l = min(hMax, vMax)
        return l * l
        
# @lc code=end

File: Answer/2943.最大化网格图中正方形空洞的面积.java
/*
 * @lc app=leetcode.cn id=2943 lang=java
 *
 * [2943] 最大化网格图中正方形空洞的面积
 */

// @lc code=start
import java.util.Arrays;

class Solution {

    public int maximizeSquareHoleArea(int n, int m, int[] hBars, int[] vBars) {
        Arrays.sort(hBars);
        var hMax = 1;
        var left = 1;
        var right = 2;
        for (var bar : hBars) {
            if (bar == right) {
                right = bar + 1;
            } else {
                left = bar - 1;
                right = bar + 1;
            }
            hMax = Math.max(hMax, right - left);
        }

        Arrays.sort(vBars);
        var vMax = 1;
        left = 1;
        right = 2;
        for (var bar : vBars) {
            if (bar == right) {
                right = bar + 1;
            } else {
                left = bar - 1;
                right = bar + 1;
            }
            vMax = Math.max(vMax, right - left);
        }

        var l = Math.min(hMax, vMax);
        return l * l;
    }
}
// @lc code=end

